# Curso de Introducción a la Ciencia de Datos

**Ejercicio SF Salaries**
Utilizaremos el `SF Salaries Dataset` Kaggle. Simplemente siga y complete las tareas que se detallan a continuación en negrita. Las tareas serán cada vez más difíciles a medida que avance.

La idea principal es aplicar pandas para organizar informacion, manipularla, buscar información en el dataset


**Importe pandas como pd. e imprima la version que esta usando**

In [111]:
# Copie el código aca

import pandas as pd

In [112]:
version_pd = pd.__version__
print("la versión de pandas es:", version_pd)

la versión de pandas es: 3.0.2


**leer el archivo Salaries.csv en un dataframe llamado salarios**

se encuentra en esta url y se puede leer directamente con pandas

https://raw.githubusercontent.com/dpalacioj/DataAI-Fundamentos-Aplicaciones/main/Analitica/ciencia_datos_con_python/tareas/datos/Salaries.csv


In [113]:
# Copie el código aca

url = "https://raw.githubusercontent.com/dpalacioj/DataAI-Fundamentos-Aplicaciones/main/Analitica/ciencia_datos_con_python/tareas/datos/Salaries.csv"
df = pd.read_csv(url)
df.head()

,Id,EmployeeName,JobTitle,BasePay,OvertimePay,OtherPay,Benefits,TotalPay,TotalPayBenefits,Year,Notes,Agency,Status
0,1,NATHANIEL FORD,GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY,167411.18,0.00,400184.25,NaN,567595.43,567595.43,2011,NaN,San Francisco,NaN
1,2,GARY JIMENEZ,CAPTAIN III (POLICE DEPARTMENT),155966.02,245131.88,137811.38,NaN,538909.28,538909.28,2011,NaN,San Francisco,NaN
2,3,ALBERT PARDINI,CAPTAIN III (POLICE DEPARTMENT),212739.13,106088.18,16452.60,NaN,335279.91,335279.91,2011,NaN,San Francisco,NaN
3,4,CHRISTOPHER CHONG,WIRE ROPE CABLE MAINTENANCE MECHANIC,77916.00,56120.71,198306.90,NaN,332343.61,332343.61,2011,NaN,San Francisco,NaN
4,5,PATRICK GARDNER,"DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)",134401.60,9737.00,182234.59,NaN,326373.19,326373.19,2011,NaN,San Francisco,NaN


**Use el método `.info()` para encontrar cuantas Columnas de atributos hay y que tipo de datos tiene cada una.**

In [114]:
# Copie el código aca
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 13 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Id                148654 non-null  int64  
 1   EmployeeName      148654 non-null  str    
 2   JobTitle          148654 non-null  str    
 3   BasePay           148045 non-null  float64
 4   OvertimePay       148650 non-null  float64
 5   OtherPay          148650 non-null  float64
 6   Benefits          112491 non-null  float64
 7   TotalPay          148654 non-null  float64
 8   TotalPayBenefits  148654 non-null  float64
 9   Year              148654 non-null  int64  
 10  Notes             0 non-null       float64
 11  Agency            148654 non-null  str    
 12  Status            0 non-null       float64
dtypes: float64(8), int64(2), str(3)
memory usage: 14.7 MB


- Segun la información del dataframe su tamaño es de *memory usage: 14.7+ MB*

- las columnas **Notes** y **Status** solamente tienen valores *NaN*, verifique que dice **0 non-null** osea 0 valores son diferentes a *NaN*
entonces todos los valores son *NaN* en esas columnas

**Elimine las columnas *Notes* y *Status* Recuerde usar el método de forma permanente (inplace), y luego visualice cuantas filas y columnas tiene luego de eliminar las columnas**

In [115]:
# Copie el código aca

df.drop(["Notes","Status"], axis=1, inplace=True)


**¿Cual es la nueva información del dataframe?**

In [116]:
df.shape
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 11 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Id                148654 non-null  int64  
 1   EmployeeName      148654 non-null  str    
 2   JobTitle          148654 non-null  str    
 3   BasePay           148045 non-null  float64
 4   OvertimePay       148650 non-null  float64
 5   OtherPay          148650 non-null  float64
 6   Benefits          112491 non-null  float64
 7   TotalPay          148654 non-null  float64
 8   TotalPayBenefits  148654 non-null  float64
 9   Year              148654 non-null  int64  
 10  Agency            148654 non-null  str    
dtypes: float64(6), int64(2), str(3)
memory usage: 12.5 MB


- Se puede verificar que al borrar las columnas que no tenian solo *NaN* se disminuye el tamaño del dataframe a: *memory usage: 12.5+ MB*

**Observe los primeros 5 registros del DataFrame.**

In [117]:
# Copie el código aca
df.head()

,Id,EmployeeName,JobTitle,BasePay,OvertimePay,OtherPay,Benefits,TotalPay,TotalPayBenefits,Year,Agency
0,1,NATHANIEL FORD,GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY,167411.18,0.00,400184.25,NaN,567595.43,567595.43,2011,San Francisco
1,2,GARY JIMENEZ,CAPTAIN III (POLICE DEPARTMENT),155966.02,245131.88,137811.38,NaN,538909.28,538909.28,2011,San Francisco
2,3,ALBERT PARDINI,CAPTAIN III (POLICE DEPARTMENT),212739.13,106088.18,16452.60,NaN,335279.91,335279.91,2011,San Francisco
3,4,CHRISTOPHER CHONG,WIRE ROPE CABLE MAINTENANCE MECHANIC,77916.00,56120.71,198306.90,NaN,332343.61,332343.61,2011,San Francisco
4,5,PATRICK GARDNER,"DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)",134401.60,9737.00,182234.59,NaN,326373.19,326373.19,2011,San Francisco


**Observe los ultimos 5 registros del DataFrame.**

In [118]:
# Copie el código aca
df.tail()

,Id,EmployeeName,JobTitle,BasePay,OvertimePay,OtherPay,Benefits,TotalPay,TotalPayBenefits,Year,Agency
148649,148650,Roy I Tillery,Custodian,0.0,0.0,0.00,0.0,0.00,0.00,2014,San Francisco
148650,148651,Not provided,Not provided,NaN,NaN,NaN,NaN,0.00,0.00,2014,San Francisco
148651,148652,Not provided,Not provided,NaN,NaN,NaN,NaN,0.00,0.00,2014,San Francisco
148652,148653,Not provided,Not provided,NaN,NaN,NaN,NaN,0.00,0.00,2014,San Francisco
148653,148654,Joe Lopez,"Counselor, Log Cabin Ranch",0.0,0.0,-618.13,0.0,-618.13,-618.13,2014,San Francisco


**¿Cuantas filas y columnas tiene el dataframe?**

In [119]:
# Copie el código aca
print("El DataFrame tiene", df.shape[0], "filas y", df.shape[1], "columnas")

El DataFrame tiene 148654 filas y 11 columnas


**Observe los 10 registros que estan desde el indice 3000**

In [120]:
# Copie el código aca
df.loc[3000:3010]

,Id,EmployeeName,JobTitle,BasePay,OvertimePay,OtherPay,Benefits,TotalPay,TotalPayBenefits,Year,Agency
3000,3001,WILLIAM BOYLE,FIREFIGHTER,105934.65,18530.98,16389.60,NaN,140855.23,140855.23,2011,San Francisco
3001,3002,ALAN LIN,ELECTRONIC MAINTENANCE TECHNICIAN,98965.80,41504.40,380.00,NaN,140850.20,140850.20,2011,San Francisco
3002,3003,WILLIAM CONLEY,SERGEANT III (POLICE DEPARTMENT),130457.78,3131.42,7256.39,NaN,140845.59,140845.59,2011,San Francisco
3003,3004,ROBERT MCDONALD,"LIEUTENANT, FIRE DEPARTMENT",123104.99,1833.56,15897.68,NaN,140836.23,140836.23,2011,San Francisco
3004,3005,MICHAEL COSTELLO,POLICE OFFICER III,112401.08,7957.74,20466.93,NaN,140825.75,140825.75,2011,San Francisco
3005,3006,EDNA HO,MANAGER V,140821.37,0.00,0.00,NaN,140821.37,140821.37,2011,San Francisco
3006,3007,RODRIGO CONEL,REGISTERED NURSE,126120.00,3981.78,10714.21,NaN,140815.99,140815.99,2011,San Francisco
3007,3008,VICTOR HURTADO,FIREFIGHTER,105934.67,13112.76,21763.50,NaN,140810.93,140810.93,2011,San Francisco
3008,3009,DONALD LANDRY,TRANSIT POWER LINE WORKER,92861.50,39789.32,8135.86,NaN,140786.68,140786.68,2011,San Francisco
3009,3010,FRANK LEE,"INSPECTOR III, (POLICE DEPARTMENT)",130481.56,725.40,9579.58,NaN,140786.54,140786.54,2011,San Francisco



La columna *Benefits* tiene valores `NaN`, ¿Cuantos NaN tiene la columna *Benefits*? (TIP: Seleccione la Columna y vefique cuantos Na tiene `.isna()` luego sumelos `.sum()` )

In [121]:
# Copie el código aca
print("La columna 'Benefits' tiene", df[["Benefits"]].isna().sum())

La columna 'Benefits' tiene Benefits    36163
dtype: int64


Al parecer los valores faltantes en la columna Benefits es por que no se le pagan beneficios entonces:

**Reemplace los valores *NaN* de la Columna *Benefits* por el valor de cero (Recuerde usar el método de forma inplace) y mostrar los primeros 10 elementos de la lista**

In [122]:
# Copie el código aca
df["Benefits"] = df["Benefits"].fillna(0)
print(df["Benefits"].loc[148645:148653])

148645    0.0
148646    0.0
148647    0.0
148648    0.0
148649    0.0
148650    0.0
148651    0.0
148652    0.0
148653    0.0
Name: Benefits, dtype: float64


**¿Cual es la nueva información del dataframe?**

In [124]:
# Copie el código aca
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 11 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   Id                148654 non-null  int64  
 1   EmployeeName      148654 non-null  str    
 2   JobTitle          148654 non-null  str    
 3   BasePay           148045 non-null  float64
 4   OvertimePay       148650 non-null  float64
 5   OtherPay          148650 non-null  float64
 6   Benefits          148654 non-null  float64
 7   TotalPay          148654 non-null  float64
 8   TotalPayBenefits  148654 non-null  float64
 9   Year              148654 non-null  int64  
 10  Agency            148654 non-null  str    
dtypes: float64(6), int64(2), str(3)
memory usage: 12.5 MB


**Cuantos valores unicos hay en las columnas categoricas?**

In [87]:
# Copie el código aca

- la columna **Agency** solamente tiene un valor unico, entonces no aporta informacion, por lo tanto se puede eliminar

**Elimine la columna *Agency* del dataframe**

In [88]:
# Copie el código aca

**Cual es el promedio de BasePay ?**

In [89]:
# Copie el código aca

**Cual es la cantidad mas alta de OvertimePay en el dataset ?**

In [90]:
# Copie el código aca

**Realice una descripcion estadistica de las columnas numericas?**

In [91]:
# Copie el código aca

**En este dataset las variables `Id` y `Year` representan variables categorica, convertir estas columnas a su formato correcto**

In [92]:
# Copie el código aca

**la variable `Year` es categorica ordinal, primero detecte cuales son los valores unicos que tiene la columna**

In [93]:
# Copie el código aca

**Apartir de la información anterior convierta la columna `Year` en variable categorica ordinal y visualice la columna**

In [94]:
# Copie el código aca

**Realice nuevamente una descripcion estadistica de las columnas numericas?**

In [95]:
# Copie el código aca

**¿Cuál es el título de trabajo de JOSEPH DRISCOLL? Nota: Utilice mayúsculas, de lo contrario, puede obtener una respuesta que no coincida.**

In [96]:
# Copie el código aca

**¿Cuanto gana JOSEPH DRISCOLL (incluyendo beneficios)?**

In [97]:
# Copie el código aca

**¿Cual es el nombre de la persona con mejor pago (incluyendo beneficios)?**

In [98]:
# Copie el código aca

**¿Cual es el nombre de la persona peor paga (incluyendo beneficios)? ¿Notan algo extraño sobre cuánto le pagan?**

In [99]:
# Copie el código aca

**Cuantos registros tienen un valor de TotalPayBeneficts menor que cero**

In [100]:
# Copie el código aca

**¿Cual es el promedio (mean) BasePay de todos los empleados por año? los años van entre (2011-2014)?** (Tip: Usar groupby)

In [101]:
# Copie el código aca

**¿Cuantos titulos de trabajos (JobTitle) unicos existen?**

In [102]:
# Copie el código aca

**¿Cuales son los 5 trabajos mas comunes?**

In [103]:
# Copie el código aca

**¿Cuántos puestos de trabajo Unicos existieron en 2013? (Ejemplo, los títulos de trabajo con solo una ocurrencia en 2013?)**

In [104]:
# Copie el código aca

**¿Cuántas personas tienen la palabra Chief en el título de su trabajo?** Apoyese en los métodos para manipular strings y en el método `.apply()`

In [105]:
# Copie el código aca

Este código es valido, se recomienda hacer el código mas simple y agregar typing y documentación.

**Bono: ¿Existe una correlación entre la longitud del nombre del Título del trabajo y el Salario?**

crea una nueva columna donde se guarde la longitud del titulo del trabajo. tip: usar el método `.apply()`

In [106]:
# Copie el código aca

Utilice el método `.corr()` de Pandas para calcular la correlacion

In [107]:
# Copie el código aca